### ISSUE TITLE
Refactor loading state

### ISSUE THREAD

**Issue by msecret** [18F/cg-dashboard#528]
In order for developers to easily add new information to the app without having to create new patterns AND for dashboard users to see correct information on when data on an app is loading or there is a lack of information, there should be set patterns for loading state.

**Comment by msecret** [comment]
Loading state should accommodate the following situations:
- Show a loading signal on first page load when no data exists and it is unknown whether there is data or not.
- Show a loading state AND the data when there is data and a request has gone out to get the newest data. This would occur when visiting a page you previously visited.
- When data was loaded successfully but there is a lack of data (yea have no apps or instances or whatever)
- Loading state for different parts of data when a single page has multiple sources of data to fetch.

**Comment by msecret** [comment]
In order to make this work, I had some ideas:

### For bullet points 4
#### Idea 1: Add loading state bool to each part of data in the object

Example:
```js
class AppStore {
  data = [
    {
      guid: 'guid-one',
      name: 'app one',
      loading: false,
      stats: {
        loading: true
      }
    }
  ]

  on FETCH_APP:
    this.data.push({ guid: action.guid, fetching: true });

  on FETCH_APP_STATS:
    const app = this.get(action.guid);
    app = assign(app, { fetching: true });

  on RESV_APP:
    const app = this.get(action.app.guid);
    app = assign(action.app, { fetching: false });

  on RESV_APP_STATS:
    const app = this.get(action.app.guid);
    app = assign(action.app.stats, {fetching: false});
}

UserStore
// This one doesn't work because it's the same user object that gets shared between
// the org and the space, so you can't say your fetching on or the other.
data = [
  {
    guid: 'guid-one',
    name: 'user one'
    orgGuid: 'guid-org-one'
  }
];
```

#### Idea 1: Stores can track specifics on what is being fetched

```js
class AppStore {
  constructor() {
    this.appFetching = false;
    this.statsFetching = false;
  }

  on APP_FETCH:
    this.appFetching = true;

  on APP_STATS_FETCH:
    this.appFetchingStats = true;

  on APP_RESV:
    this.appFetching = false;
}

class UserStore {
  ...

  on ORG_USERS_FETCH:
    this.orgFetching = true;

  on SPACE_USERS_FETCH:
    this.spaceFetching = true;

  on ORG_USERS_RESV:
    this.orgFetching = false;

  ...
}

```

**Comment by msecret** [comment]
#### Idea 4: track loading and empty separately

Component algorithm:
- If loading, always show a loading signal at top of screen.
- If there's no data, don't show the table yet
- If the data is verified empty (which the store is responsible for), show empty whether or not loading is true or whether there is data.
```js

class AppContainer {
  render() {
    content = <div></div>;
    loading = <div></div>
    if (AppStore.appFetching) {
      loading = <Loading "App loading" />;
    }

    if (AppStore.isEmpty()) {
      content = <h4>None</h4>
    } else if (this.state.app) {
       content = <Table apps />
    }

    return (
      <loading>
      </content>
    )
  }
}
```

Store algorithm:
- If an app hasn't been fetched yet, it isn't considered empty but app store is fetching
- If an app was fetched but theres nothing there (or there's a 404) then app store says it's empty. This could potentially just be be one empty on the whole store? We might have to separate for org vs space on the user store.
- If an app was resv and there's data, then both fetching and empty are false.

```js
class AppStore {
  constructor() {
    this.appFetching = false;
    this.appEmpty = false; // We might need this to be on the app based on guid
  }

  on APP_FETCH:
    this.appFetching = true;

  on APP_RESV:
    this.appFetching = false;
    if (isEmpty(action.app)) this.appEmpty = true;
    data.put(action.app)
}
```

**Comment by msecret** [comment]
Now that I think about this more, I realize on any list views, the view isn't interested in whether a specific entity is there, it's interested whether a group of entities with the same parent is there.

```js
class UserStore {
  
  on ORG_USERS_FETCH:
    this.data.push({org_guid: action.guid, fetching: true});

  on SPACE_USERS_FETCH:
    this.data.push({space_guid: action.guid, fetching: true});
}
```

OR

```js
class UserStore {
  
  on ORG_USERS_FETCH:
    this.orgFetching.push(action.guid);

  on SPACE_USERS_FETCH:
    this.spaceFetching.push(action.guid);
}
```

A question: should stores track specifically which data is empty or loading, or can every store just have a single boolean to track whether anything on the store is loading or empty? The latter case is much easier but could lead to incorrect data, although I don't think that's the current problem.

**Comment by msecret** [comment]
I tried idea 4 here: https://github.com/18F/cg-dashboard/pull/529. It also includes a few ideas from idea 2.

**Comment by msecret** [comment]
Fixes #520

**Comment by msecret** [comment]
Fixes #522
